File: VB/Solution3.Module.Web/Controllers/DashboardRefreshController.vb
Imports System
Imports System.Globalization
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Editors
Imports DevExpress.ExpressApp.Web
Imports DevExpress.ExpressApp.Web.Editors.ASPx
Imports DevExpress.ExpressApp.Web.Templates
Imports DevExpress.ExpressApp.Web.Utils

Namespace Solution3.Module.Web.Controllers
	Public Class DashboardRefreshController
		Inherits ViewController(Of DashboardView)

		Public Const FilterSourceID As String = "FilterSource"
		Protected Overrides Sub OnViewControlsCreated()
			MyBase.OnViewControlsCreated()
			If WebWindow.CurrentRequestWindow Is Nothing Then
				Return
			End If
			RemoveHandler WebWindow.CurrentRequestWindow.PagePreRender, AddressOf CurrentRequestWindow_PagePreRender
			AddHandler WebWindow.CurrentRequestWindow.PagePreRender, AddressOf CurrentRequestWindow_PagePreRender
		End Sub

		Private Sub CurrentRequestWindow_PagePreRender(ByVal sender As Object, ByVal e As EventArgs)
			If View Is Nothing Then
				Return
			End If
			Dim sourceItem As DashboardViewItem = CType(View.FindItem(FilterSourceID), DashboardViewItem)
			If sourceItem.InnerView Is Nothing Then
				Return
			End If
			Dim listView As ListView = CType(sourceItem.InnerView, ListView)
			Dim editor As ASPxGridListEditor = CType(listView.Editor, ASPxGridListEditor)
			If editor Is Nothing Then
				Return
			End If
			Dim holder As ICallbackManagerHolder = DirectCast(WebWindow.CurrentRequestPage, ICallbackManagerHolder)
			Dim script As String = holder.CallbackManager.GetScript()
			script = String.Format(CultureInfo.InvariantCulture, "" & ControlChars.CrLf & _
"                function(s, e) {{" & ControlChars.CrLf & _
"                    var xafCallback = function() {{" & ControlChars.CrLf & _
"                        s.EndCallback.RemoveHandler(xafCallback);" & ControlChars.CrLf & _
"                        {0}" & ControlChars.CrLf & _
"                    }};" & ControlChars.CrLf & _
"                    s.EndCallback.AddHandler(xafCallback);" & ControlChars.CrLf & _
"                }}" & ControlChars.CrLf & _
"                ", script)
			ClientSideEventsHelper.AssignClientHandlerSafe(editor.Grid, "SelectionChanged", script, "DashboardRefreshController")
		End Sub
	End Class
End Namespace

File: CS/Solution3.Module.Web/Controllers/DashboardRefreshController.cs
using System;
using System.Globalization;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Editors;
using DevExpress.ExpressApp.Web;
using DevExpress.ExpressApp.Web.Editors.ASPx;
using DevExpress.ExpressApp.Web.Templates;
using DevExpress.ExpressApp.Web.Utils;

namespace Solution3.Module.Web.Controllers {
    public class DashboardRefreshController : ViewController<DashboardView> {
        public const string FilterSourceID = "FilterSource";
        protected override void OnViewControlsCreated() {
            base.OnViewControlsCreated();
            if(WebWindow.CurrentRequestWindow == null) return;
            WebWindow.CurrentRequestWindow.PagePreRender -= CurrentRequestWindow_PagePreRender;
            WebWindow.CurrentRequestWindow.PagePreRender += CurrentRequestWindow_PagePreRender;
        }

        private void CurrentRequestWindow_PagePreRender(object sender, EventArgs e) {
			if(View == null) return;
            DashboardViewItem sourceItem = (DashboardViewItem)View.FindItem(FilterSourceID);
            if(sourceItem.InnerView == null) return;
            ListView listView = (ListView)sourceItem.InnerView;
            ASPxGridListEditor editor = (ASPxGridListEditor)listView.Editor;
            if(editor == null) return;
            ICallbackManagerHolder holder = (ICallbackManagerHolder)WebWindow.CurrentRequestPage;
            string script = holder.CallbackManager.GetScript();
            script = string.Format(CultureInfo.InvariantCulture, @"
function(s, e) {{
    if(e.isChangedOnServer){{
        {0}
    }}else{{
        var xafCallback = function() {{
            s.EndCallback.RemoveHandler(xafCallback);
            {0}
        }};
        s.EndCallback.AddHandler(xafCallback);
    }}
}}
                ", script);
            ClientSideEventsHelper.AssignClientHandlerSafe(editor.Grid, "SelectionChanged", script, "DashboardRefreshController");
        }
    }
}
